### PR TITLE
Refactor folded/unfolded/overlaid/draped states

### PR DESCRIPTION
**Pull request by ChengshuLi**

### CODE REVIEW

File: omnigibson/examples/object_states/overlaid_demo.py [hunk 1]
```diff
@@ -32,7 +32,7 @@ def main(random_selection=False, headless=False, short_exec=False):
                 "category": "carpet",
                 "model": "ctclvd",
                 "prim_type": PrimType.CLOTH,
-                "abilities": {"foldable": {}},
+                "abilities": {"cloth": {}},
```
Comment (cremebrule) [inline omnigibson/examples/object_states/overlaid_demo.py:35]: nice!! so all cloth ==> implies has un/folded state right?

File: omnigibson/object_states/draped.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        """
+        Check whether the (cloth) object is draped on the other (rigid) object.
+        The cloth object needs to touch the rigid object and its velocity needs to be small.
+        """
+        if not (self.obj.prim_type == PrimType.CLOTH and other.prim_type == PrimType.RIGID):
+            raise ValueError("Draped state requires obj1 is cloth and obj2 is rigid.")
+
+        return self.obj.states[Touching].get_value(other)
+
+        # return self.obj.states[Touching].get_value(other) and \
```
Comment (cremebrule) [inline omnigibson/object_states/draped.py:55]: outdated code?

File: omnigibson/object_states/draped.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+# Create settings for this module
+m = create_module_macros(module_path=__file__)
+
+# Maximum velocity allowed for the draped object
+m.DRAPED_MAX_VELOCITY = 0.05
+
+class Draped(KinematicsMixin, RelativeObjectState, BooleanState):
+    @staticmethod
+    def get_dependencies():
+        return KinematicsMixin.get_dependencies() + RelativeObjectState.get_dependencies() + [Touching]
```
Comment (cremebrule) [inline omnigibson/object_states/draped.py:26]: nit: in general, might be safer to make this more programmatic -- e.g.: call super().get_dependencies() first and then add to that

File: omnigibson/object_states/draped.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+    def _get_value(self, other):
+        """
+        Check whether the (cloth) object is draped on the other (rigid) object.
+        The cloth object needs to touch the rigid object and its velocity needs to be small.
+        """
+        if not (self.obj.prim_type == PrimType.CLOTH and other.prim_type == PrimType.RIGID):
+            raise ValueError("Draped state requires obj1 is cloth and obj2 is rigid.")
+
+        return self.obj.states[Touching].get_value(other)
```
Comment (cremebrule) [inline omnigibson/object_states/draped.py:53]: hmmmm...is this sufficient? what if the cloth is on the ground, and touching a table's leg?

File: omnigibson/object_states/folded.py [hunk 5]
```diff
@@ -1,50 +1,150 @@
...
+# Angle threshold for checking smoothness of the cloth; surface normals need to be close enough to the z-axis
+m.NORMAL_Z_ANGLE_DIFF = np.deg2rad(45.0)
+
+"""
+FoldedLevelData contains the following fields:
+    smoothness (float): percentage of surface normals that are sufficiently close to the z-axis
+    area (float): the area of the convex hull of the projected points compared to the initial unfolded state
+    diagonal (float): the diagonal of the convex hull of the projected points compared to the initial unfolded state
+"""
+FoldedLevelData = namedtuple("FoldedLevelData", ("smoothness", "area", "diagonal"))
```
Comment (cremebrule) [inline omnigibson/object_states/folded.py:34]: Oh nice!! maybe we should do the same for ContainedParticles state (and in general, any intermediate state)?

File: omnigibson/object_states/folded.py [hunk 6]
```diff
@@ -54,3 +154,19 @@ def _set_value(self, new_value):
         raise NotImplementedError("Folded does not support set_value(True)")
 
     # We don't need to dump / load anything since the cloth objects should handle it themselves
+
+class Unfolded(AbsoluteObjectState, BooleanState):
+    def _get_value(self):
+        # Check the smoothness of the cloth
+        folded_level = self.obj.states[FoldedLevel].get_value()
+        return folded_level.smoothness >= m.NORMAL_Z_PERCENTAGE and \
```
Comment (cremebrule) [inline omnigibson/object_states/folded.py:162]: so clean!!

File: omnigibson/utils/object_state_utils.py [hunk 7]
```diff
@@ -252,97 +252,79 @@ def sample_kinematics(
...
+    state = og.sim.dump_state(serialized=False)
 
+    # Reset the cloth
+    obj.root_link.reset()
 
-def calculate_projection_area_and_diagonal(obj, dims):
-    """
-    Calculate the projection area and the diagonal length when projecting to the plane defined by the input dims
-    E.g. if dims is [0, 1], the points will be projected onto the x-y plane.
+    obj_aabb_low, obj_aabb_high = obj.states[AABB].get_value()
```
Comment (cremebrule) [inline omnigibson/utils/object_state_utils.py:281]: are these values correct? what if they're already queried before `obj.root_link.reset()` was called? maybe we should clear the cache of these values first just to be safe

File: omnigibson/utils/object_state_utils.py [hunk 8]
```diff
@@ -252,97 +252,79 @@ def sample_kinematics(
...
+    Samples the cloth object @obj on the rigid object @other
 
     Args:
-        obj (DatasetObject): Must be PrimType.CLOTH
+        obj (StatefulObject): Object whose state should be sampled. e.g.: for sampling a bed sheet on a rack,
+            @obj is the bed sheet
+        other (StatefulObject): Object who is the reference point for @obj's state. e.g.: for sampling a bed sheet
+            on a rack, @other is the rack
+        max_trials (int): Number of attempts for sampling
+        z_offset (float): Z-offset to apply to the sampled pose
```
Comment (cremebrule) [inline omnigibson/utils/object_state_utils.py:265]: add randomize_xy docstring

File: omnigibson/utils/object_state_utils.py [hunk 9]
```diff
@@ -252,97 +252,79 @@ def sample_kinematics(
...
-        diagonal (float): diagonal of the convex hull of the projected points
-    """
-    cloth = obj.links["base_link"]
-    points = cloth.keypoint_particle_positions[:, dims]
-    hull = ConvexHull(points)
+    if randomize_xy:
+        low = np.array([other_aabb_low[0], other_aabb_low[1], z_value])
+        high = np.array([other_aabb_high[0], other_aabb_high[1], z_value])
+    else:
+        low = np.array([(other_aabb_low[0] + other_aabb_high[0]) / 2.0,
```
Comment (cremebrule) [inline omnigibson/utils/object_state_utils.py:291]: maybe quick comment here? default is sampling cloth at the center of @other

File: omnigibson/utils/object_state_utils.py [hunk 10]
```diff
@@ -252,97 +252,79 @@ def sample_kinematics(
...
+        if success:
+            break
+        else:
+            og.sim.load_state(state)
 
-    if m.DEBUG_CLOTH_PROJ_VIS:
-        import matplotlib.pyplot as plt
-        ax = plt.gca()
-        ax.set_aspect('equal')
+    if success:
```
Comment (cremebrule) [inline omnigibson/utils/object_state_utils.py:314]: why is this needed? isn't this already immediately done prior anyways?

Comment (cremebrule) [reply omnigibson/utils/object_state_utils.py:314]: (the obj.set_position_orientation(), reset(), and keep_still() calls)

File: tests/test_object_states.py [hunk 11]
```diff
@@ -756,5 +776,125 @@ def test_attached_to():
...
+    assert not bottom_cabinet.states[Open].get_value()
+
+    # Set the joints to their upper limits.
+    microwave.joints["j_link_0"].set_pos(microwave.joints["j_link_0"].upper_limit)
+    bottom_cabinet.joints["j_link_2"].set_pos(bottom_cabinet.joints["j_link_2"].upper_limit)
+
+    og.sim.step()
+
+    # The objects should be open.
+    assert microwave.states[Open].get_value()
```
Comment (cremebrule) [inline tests/test_object_states.py:841]: will the open / close tertiary state be handled in a separate PR?

File: tests/test_object_states.py [hunk 12]
```diff
@@ -756,5 +776,125 @@ def test_attached_to():
...
+    assert microwave.states[Open].set_value(False)
+    assert bottom_cabinet.states[Open].set_value(False)
+
+    # The objects should not be open.
+    assert not microwave.states[Open].get_value()
+    assert not bottom_cabinet.states[Open].get_value()
+
+@og_test
+def test_draped():
+    from IPython import embed
```
Comment (cremebrule) [inline tests/test_object_states.py:872]: remove

File: tests/test_object_states.py [hunk 13]
```diff
@@ -756,5 +776,125 @@ def test_attached_to():
...
+    print("test")
+    embed()
```
Comment (cremebrule) [inline tests/test_object_states.py:874]: remove

**Review by cremebrule — COMMENTED** [review]
awesome work!! left a few comments

File: omnigibson/examples/object_states/overlaid_demo.py (hunk 1 above)
Comment (ChengshuLi) [reply omnigibson/examples/object_states/overlaid_demo.py:35]: I think so.

File: omnigibson/object_states/draped.py (hunk 3 above)
Comment (ChengshuLi) [reply omnigibson/object_states/draped.py:26]: I don't think I understand this fully. Let's do a proper cleanup in the next PR.

**Review by cremebrule — APPROVED** [review]
LGTM!
